### PR TITLE
Calibrated the Regressor Model

### PR DESCRIPTION
**Pull request by WhiteWolf47**
This PR aims to calibrate the Regressor model using the IsotonicRegressionCalibrator class from the model_calibration.py file.

### CODE REVIEW

**Review by suhaibmujahid — COMMENTED** [review]
Thank you, @WhiteWolf47! Could you please share the performance reports here (for the before and after)?

File: bugbug/models/regressor.py [hunk 1]
```diff
@@ -128,8 +129,12 @@ def __init__(
             ]
         )
 
-        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
-        self.clf.set_params(predictor="cpu_predictor")
+        base_clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+        base_clf.set_params(predictor="cpu_predictor")
+        self.clf = IsotonicRegressionCalibrator(base_clf)
+
+        # this is a temporary workaround for now
```
[the remark this answers is not in the document]
Comment (WhiteWolf47) [reply bugbug/models/regressor.py:136]: okay

**Comment by WhiteWolf47** [comment]
@suhaibmujahid here are the reports -

before calibration - 
[output.txt](https://github.com/mozilla/bugbug/files/11312061/output.txt)

after calibration - 
[output_cal.txt](https://github.com/mozilla/bugbug/files/11312066/output_cal.txt)

File: bugbug/models/regressor.py [hunk 2]
```diff
@@ -128,8 +129,12 @@ def __init__(
             ]
         )
 
-        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
-        self.clf.set_params(predictor="cpu_predictor")
+        base_clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+        base_clf.set_params(predictor="cpu_predictor")
+        self.clf = IsotonicRegressionCalibrator(base_clf)
+
+        # this is a temporary workaround for the error - "shap.utils._exceptions.InvalidModelError: Model type not yet supported by TreeExplainer: <class 'bugbug.model_calibration.IsotonicRegressionCalibrator'>"
```
Comment (suhaibmujahid) [inline bugbug/models/regressor.py:136]:
```suggestion
        # This is a temporary workaround for the error: "Model type not yet supported by TreeExplainer"
```

**Review by suhaibmujahid — APPROVED** [review]
Thank you, LGTM!

File: bugbug/models/regressor.py [hunk 3]
```diff
@@ -128,8 +129,12 @@ def __init__(
             ]
         )
 
-        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
-        self.clf.set_params(predictor="cpu_predictor")
+        base_clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+        base_clf.set_params(predictor="cpu_predictor")
+        self.clf = IsotonicRegressionCalibrator(base_clf)
```
Comment (marco-c) [inline bugbug/models/regressor.py:134]: Could you add a parameter to the constructor to enable/disable the calibration?

Comment (WhiteWolf47) [reply bugbug/models/regressor.py]: @marco-c done!

Comment (marco-c) [reply bugbug/models/regressor.py]: @WhiteWolf47 what you added isn't a parameter, but an attribute.
A parameter is a variable that is passed to the function (see for example `interpretable`).

File: bugbug/models/regressor.py [hunk 4]
```diff
@@ -127,9 +128,16 @@ def __init__(
                 ("union", ColumnTransformer(column_transformers)),
             ]
         )
-
-        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
-        self.clf.set_params(predictor="cpu_predictor")
+        self.calibration = True
+        if self.calibration:
+            base_clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+            base_clf.set_params(predictor="cpu_predictor")
```
Comment (suhaibmujahid) [inline bugbug/models/regressor.py:134]: This common logic could be done outside the calibration condition.

Comment (WhiteWolf47) [reply bugbug/models/regressor.py:134]: makes sense!

File: bugbug/models/regressor.py (hunk 3 above)
Comment (WhiteWolf47) [reply bugbug/models/regressor.py]: mb :sweat_smile: , i misunderstood it!

**Review by suhaibmujahid — DISMISSED** [review]

File: bugbug/models/regressor.py [hunk 5]
```diff
@@ -127,9 +129,14 @@ def __init__(
...
-        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
-        self.clf.set_params(predictor="cpu_predictor")
+        base_clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
+        base_clf.set_params(predictor="cpu_predictor")
+        if calibration:
+            self.clf = IsotonicRegressionCalibrator(base_clf)
+            # This is a temporary workaround for the error : "Model type not yet supported by TreeExplainer"
+            self.calculate_importance = False
+        else:
+            self.clf = base_clf
```
Comment (marco-c) [inline bugbug/models/regressor.py:139]: Nit: You could simplify this code by doing:
```suggestion
        self.clf = xgboost.XGBClassifier(n_jobs=utils.get_physical_cpu_count())
        self.clf.set_params(predictor="cpu_predictor")

        if calibration:
            self.clf = IsotonicRegressionCalibrator(self.clf)
            # This is a temporary workaround for the error : "Model type not yet supported by TreeExplainer"
            self.calculate_importance = False
```

**Review by marco-c — DISMISSED** [review]

**Review by suhaibmujahid — APPROVED** [review]

**Pull request merged**